### PR TITLE
[Fix] Fix cpu inference UT failure

### PR DESCRIPTION
**Pull request by delock**
This PR fix UT test error as described in this PR and the following test job.  This PR skips `TestModelTask` if dtype is not supported by accelerator, or `InferenceBuilder` is not implemented by accelerator.
https://github.com/microsoft/DeepSpeed/pull/4419
https://github.com/microsoft/DeepSpeed/actions/runs/6341645987/job/17235544538

### CODE REVIEW

**Review by loadams — APPROVED** [review]

**Comment by loadams** [comment]
FYI @delock - this looks like it only runs the same 4 tests that it did before?  Just making sure that's intentional.  Also because of the new placement of the skips rather than the whole file, the tests run much slower than before.  We will look at fixing that, but wanted sure make sure that the total number of tests should be higher.

**Comment by delock** [comment]
Hi @loadams , currently we are seeking higher test coverage in the area of AutoTP.  @Yejing-Lai is currently investigating whether more model coverage is possible.   On the other hand.  Some tests skipped because on CPU `InferenceBuilder` are not implemented.  We will check if there is any skipped test not because of this reason.

> FYI @delock - this looks like it only runs the same 4 tests that it did before? Just making sure that's intentional. Also because of the new placement of the skips rather than the whole file, the tests run much slower than before. We will look at fixing that, but wanted sure make sure that the total number of tests should be higher.

**Comment by loadams** [comment]
> Hi @loadams , currently we are seeking higher test coverage in the area of AutoTP. @Yejing-Lai is currently investigating whether more model coverage is possible. On the other hand. Some tests skipped because on CPU `InferenceBuilder` are not implemented. We will check if there is any skipped test not because of this reason.
> 
> > FYI @delock - this looks like it only runs the same 4 tests that it did before? Just making sure that's intentional. Also because of the new placement of the skips rather than the whole file, the tests run much slower than before. We will look at fixing that, but wanted sure make sure that the total number of tests should be higher.

That makes sense, thanks.  Given that this increases the runtime of the cpu_inference tests drastically without increasing coverage, do you think it would make sense to either revert it or add the skip back to the top of the file which seemed to run quicker, then on subsequent PRs we can enable more tests and monitor the overall runtime?

**Comment by loadams** [comment]
> > Hi @loadams , currently we are seeking higher test coverage in the area of AutoTP. @Yejing-Lai is currently investigating whether more model coverage is possible. On the other hand. Some tests skipped because on CPU `InferenceBuilder` are not implemented. We will check if there is any skipped test not because of this reason.
> > > FYI @delock - this looks like it only runs the same 4 tests that it did before? Just making sure that's intentional. Also because of the new placement of the skips rather than the whole file, the tests run much slower than before. We will look at fixing that, but wanted sure make sure that the total number of tests should be higher.
> 
> That makes sense, thanks. Given that this increases the runtime of the cpu_inference tests drastically without increasing coverage, do you think it would make sense to either revert it or add the skip back to the top of the file which seemed to run quicker, then on subsequent PRs we can enable more tests and monitor the overall runtime?

FYI @delock, we disabled the cpu_inference test, but have it so we can run it on demand for PRs that we think will impact things as we fix the test time issue

**Comment by delock** [comment]
I tried to run the line: `pytest -m 'seq_inference' unit/` locally and see 3 passed  tests with this branch.  Yet the same branch get these three test skipped on CI environment.  Will need some further investigation with the CI environment.
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-t5] PASSED                                                                                                                                                                          [ 66%]
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-roberta] PASSED                                                                                                                                                                     [ 73%] unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-marian] SKIPPED (Acceleraor cpu does not support torch.float16.)                                                                                                              [ 80%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-codegen] SKIPPED (Acceleraor cpu does not support torch.float16.)                                                                                                             [ 86%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[bf16-marian] PASSED
> FYI @delock - this looks like it only runs the same 4 tests that it did before? Just making sure that's intentional. Also because of the new placement of the skips rather than the whole file, the tests run much slower than before. We will look at fixing that, but wanted sure make sure that the total number of tests should be higher.

**Comment by delock** [comment]
The slow running of test is caused by this line.   It loads model list from huggingface_hub.  Doing it for every test is slow.   A proper fix should make it persistent during the test session.  One way is save it as pickle in the first test and load the pickle subsequently.   Will test and update this PR.
https://github.com/microsoft/DeepSpeed/blob/master/tests/unit/inference/test_inference.py#L68

**Comment by delock** [comment]
@loadams the slow test had been fixed with latest commit.   How the test run around 13 minutes on my local environment.

The strange thing is on my local environment, there are three tests not skipped, and the skipping message is displayed in my local environment .

On github CI environment, however, these three tests are skipp and the skipping message is not printed with the test report.
I tried several attempt to turn on skip message printing, however it seems pytest in CI environment are not affected  by the pytest command in the workflow.  Do you have any idea and how do I turn on skip message printing in github CI env?

This is my local env.
```
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-neo] SKIPPED (This op had not been implemented on this...) [  6%]
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-neox] SKIPPED (Skipping gpt-neox-20b for now)              [ 13%]
unit/inference/test_inference.py::TestMPSize::test[fp32-bloom] SKIPPED (Bloom models only support half precision, ...) [ 20%]
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-j] SKIPPED (Test is currently broken)                      [ 26%]
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-neo] SKIPPED (This op had not been implemented on this...) [ 33%]
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-neox] SKIPPED (Skipping gpt-neox-20b for now)              [ 40%]
unit/inference/test_inference.py::TestMPSize::test[fp16-bloom] SKIPPED (This op had not been implemented on this s...) [ 46%]
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-j] SKIPPED (Test is currently broken)                      [ 53%]
unit/inference/test_inference.py::TestAutoTP::test[falcon] SKIPPED (Not enough GPU memory for this on V100 runners)    [ 60%]
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-t5] PASSED                                            [ 66%]
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-roberta] PASSED                                       [ 73%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-marian] SKIPPED (Acceleraor cpu does not su...) [ 80%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-codegen] SKIPPED (Acceleraor cpu does not s...) [ 86%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[bf16-marian] PASSED                                  [ 93%]
unit/inference/test_inference.py::TestAutoTensorParallelism::test[bf16-codegen] SKIPPED (Codegen model(bf16) need ...) [100%]
```

This is the github CI printout, all tests are skipped and no skip message printed.
```
unit/inference/test_inference.py::TestAutoTensorParallelism::test[bf16-marian] SKIPPED
unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-marian] SKIPPED
unit/inference/test_inference.py::TestAutoTensorParallelism::test[fp16-codegen] SKIPPED
unit/inference/test_inference.py::TestAutoTensorParallelism::test[bf16-codegen] SKIPPED
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-roberta] SKIPPED
unit/inference/test_inference.py::TestInjectionPolicy::test[fp32-t5] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp16-bloom] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp32-bloom] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-j] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-neox] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-neo] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp32-gpt-j] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-neo] SKIPPED
unit/inference/test_inference.py::TestMPSize::test[fp16-gpt-neox] SKIPPED
unit/inference/test_inference.py::TestAutoTP::test[falcon] SKIPPED (...)\
```

**Comment by delock** [comment]
The skip message is truncated because of 80 column limit when pytest is running.  We can expand the 'COLUMNS' env variable to 140 to see skip message.   Another issue is recently pytorch upgraded to 2.1.0, this caused Intel Extension for PyTorch load error which subsequently caused all test skipped.   Will discuss with Intel Extension for PyTorch team when they will release a wheel for PyTorch 2.1.0.

**Comment by delock** [comment]
Intel Extension for PyTorch 2.1 is released.  Will update this PR to change workflow to pytorch 2.1 accordingly.

**Comment by delock** [comment]
Hi @loadams  @tjruwase, we are adding back tensor parallel UT into CPU inference workflow.  One issue we met is github instance "ubuntu-20.04" has only one CPU and two cores.   We need a stronger system if we want to run autotp test stablely.  We will use one of the self-hosted v100.  Can you help start the workflow?  Thanks!

File: .github/workflows/cpu-inference.yml
Comment (loadams) [inline .github/workflows/cpu-inference.yml:18]: FYI @mrwyattii - see the comment on the PR for the explanation of why.

Comment (mrwyattii) [reply .github/workflows/cpu-inference.yml:18]: @loadams @delock we have self-hosted CPU runners. Can we please use those instead to avoid wasting GPU resources?

Use `runs-on: [self-hosted, cpu]`

Comment (delock) [reply .github/workflows/cpu-inference.yml:18]: Thanks for the suggestion, updated.

**Comment by delock** [comment]
Note currently there is a CPU test failure due to compatibility between https://github.com/microsoft/DeepSpeed/commit/4fc181b01077521ba42379013ce91a1c294e5d8e and pytorch 2.1.  We are working on a fix.   Will go back to you when a PR is ready

**Comment by delock** [comment]
This failure should be fixed by https://github.com/microsoft/DeepSpeed/pull/4578.   We should merge #4578 into this PR so they can be tested together on CPU host
> Note currently there is a CPU test failure due to compatibility between [4fc181b](https://github.com/microsoft/DeepSpeed/commit/4fc181b01077521ba42379013ce91a1c294e5d8e) and pytorch 2.1. We are working on a fix. Will go back to you when a PR is ready

**Comment by delock** [comment]
#4578 is merged and CPU workflow ad-hoc running is turned off.  @loadams can you help start the workflow to see whether cpu_inference workflow passes?  Thanks!

**Comment by loadams** [comment]
> #4578 is merged and CPU workflow ad-hoc running is turned off. @loadams can you help start the workflow to see whether cpu_inference workflow passes? Thanks!

@delock - I tried cherrypicking these changes into a branch of mine so that I could run the workflow dispatch.  It was run here, but I'm seeing errors still: https://github.com/microsoft/DeepSpeed/actions/runs/6709501151/job/18232686468

In this PR, I think you can re-enable the PR build flow and it should run here now.

File: .github/workflows/cpu-inference.yml
Comment (loadams) [inline .github/workflows/cpu-inference.yml:86]: Might be worth moving these up one section after the call to `pip list`?

**Comment by delock** [comment]
Hi @loadams , thanks for help on the workflow.  From this line it seems gcc on self-hosted cpu node is using gcc-7, which have compatibility problem with AVX intrinsics in ccl.cpp kernels.
https://github.com/microsoft/DeepSpeed/actions/runs/6709501151/job/18232686468#step:11:147
I can also get the same message from this line, gcc is 7.5.0
https://github.com/microsoft/DeepSpeed/actions/runs/6710829047/job/18236865881?pr=4430#step:7:33

From a previous github hosted CPU system, the comm ops can be successfully built.  The gcc version is 9.4.0
https://github.com/delock/DeepSpeedSYCLSupport/actions/runs/6634978766/job/18025197269#step:7:33

So I think to resolve this issue gcc needs to be switched to 9.4.0 or above.  What is the steps to switch gcc version on self-hosted CPU node?
> > #4578 is merged and CPU workflow ad-hoc running is turned off. @loadams can you help start the workflow to see whether cpu_inference workflow passes? Thanks!
> 
> @delock - I tried cherrypicking these changes into a branch of mine so that I could run the workflow dispatch. It was run here, but I'm seeing errors still: https://github.com/microsoft/DeepSpeed/actions/runs/6709501151/job/18232686468
> 
> I was able to "resolve" the merge conflict and enable this back in your PR, so it should be running below and we can test just by pushing to this PR.

**Comment by delock** [comment]
Hi @loadams  does it make sense to install gcc-9 in the workflow steps, following the link below?

https://askubuntu.com/questions/1140183/install-gcc-9-on-ubuntu-18-04 

> Hi @delock - thanks for the clarification, we don't have an easy way to update it, I will look into getting a newer image running on this, since the fact that it is running ubuntu 18.04 is part of the problem here.

**Comment by loadams** [comment]
> Hi @loadams does it make sense to install gcc-9 in the workflow steps, following the link below?
> 
> https://askubuntu.com/questions/1140183/install-gcc-9-on-ubuntu-18-04
> 
> > Hi @delock - thanks for the clarification, we don't have an easy way to update it, I will look into getting a newer image running on this, since the fact that it is running ubuntu 18.04 is part of the problem here.

That's a fine option to try for now, I'm not sure if that will work in these pipelines, I've had some trouble with it manually, but it is worth a shot here since it might unblock us while we try to get newer images on these VMs.

File: .github/workflows/cpu-inference.yml
Comment (delock) [reply .github/workflows/cpu-inference.yml:94]: These checks are moved up to the `pip list` section

**Comment by delock** [comment]
@loadams I added a section to install gcc-9, let's see if this workaround works.

**Comment by delock** [comment]
Let me find a system with ubunt-18 and figure out how this works.

> The install looks to have been successful, however the following check is still returning 7.5.0 for some reason: https://github.com/microsoft/DeepSpeed/actions/runs/6742349658/job/18341364878?pr=4430

**Comment by delock** [comment]
> Let me find a system with ubunt-18 and figure out how this works.
> 
> > The install looks to have been successful, however the following check is still returning 7.5.0 for some reason: https://github.com/microsoft/DeepSpeed/actions/runs/6742349658/job/18341364878?pr=4430

I found a ubuntu machine and did some experiment, turns how have to use the following commands to actually 'switch' to the new compiler
```
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 99
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 99
```

**Comment by loadams** [comment]
> > Let me find a system with ubunt-18 and figure out how this works.
> > > The install looks to have been successful, however the following check is still returning 7.5.0 for some reason: https://github.com/microsoft/DeepSpeed/actions/runs/6742349658/job/18341364878?pr=4430
> 
> I found a ubuntu machine and did some experiment, turns out have to use the following commands to actually 'switch' to the new compiler
> 
> ```
> sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 99
> sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 99
> ```

Looks like at least a new failure?

**Comment by delock** [comment]
> > > Let me find a system with ubunt-18 and figure out how this works.
> > > > The install looks to have been successful, however the following check is still returning 7.5.0 for some reason: https://github.com/microsoft/DeepSpeed/actions/runs/6742349658/job/18341364878?pr=4430
> > 
> > 
> > I found a ubuntu machine and did some experiment, turns out have to use the following commands to actually 'switch' to the new compiler
> > ```
> > sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 99
> > sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 99
> > ```
> 
> Looks like at least a new failure?

Yes, I'm checking with Intel Extension for Pytorch engineers for this new failure.
```
Looking in links: https://developer.intel.com/ipex-whl-stable-cpu
ERROR: Could not find a version that satisfies the requirement oneccl_bind_pt (from versions: none)
ERROR: No matching distribution found for oneccl_bind_pt
Error: Process completed with exit code 1.
```
I'm going to push an alternative way to install from the suggestion.

**Comment by delock** [comment]
@loadams We are now using an alternative way to install oneccl_bind_pt following @jingxu10 suggestion.   Also added a curl line to see if there are any connectivity issue.

**Comment by delock** [comment]
@loadams the cpu_inference workflow failed without specific information, is it because the self-hosted CPU node is not available?

**Comment by loadams** [comment]
> @loadams the cpu_inference workflow failed without specific information, is it because the self-hosted CPU node is not available?

That's what it appears to be - I re-merged the master branch to see if the test would run again and it doesn't appear to be running yet, I will monitor it and see.

**Comment by delock** [comment]
It looks like self-hosted CPU instance are not available for pickup.  Should we try us one of the self-hosted GPU instance to validate the workflow first?

**Comment by loadams** [comment]
> It looks like self-hosted CPU instance are not available for pickup. Should we try us one of the self-hosted GPU instance to validate the workflow first?

@delock - I think this is fixed now, I see it running at least.  So please push any other changes, but if we hit this again we can easily change to test on the gpu runners too

**Comment by delock** [comment]
> > It looks like self-hosted CPU instance are not available for pickup. Should we try us one of the self-hosted GPU instance to validate the workflow first?
> 
> @delock - I think this is fixed now, I see it running at least. So please push any other changes, but if we hit this again we can easily change to test on the gpu runners too

Thanks @loadams  Now I see there is a new error, I'll check with oneCCL team on this error.
https://github.com/microsoft/DeepSpeed/actions/runs/6882177085/job/18720136373?pr=4430#step:12:251

**Comment by delock** [comment]
@loadams mitigation for this error added.  Can you help start the workflow?  Thanks!
> > It looks like self-hosted CPU instance are not available for pickup. Should we try us one of the self-hosted GPU instance to validate the workflow first?
> 
> @delock - I think this is fixed now, I see it running at least. So please push any other changes, but if we hit this again we can easily change to test on the gpu runners too

**Comment by delock** [comment]
The error message shows conda package needs to be installed with sudo  in order to get permission to install into /opt/conda.  The latest CI fixed this error.  @loadams can you help start the workflow again?  Thanks!

**Comment by delock** [comment]
Hi @loadams, the permission error had been fixed.  I also merged a fix on inference_all_reduce that impacts inference_all_reduce performance on CPU.  Can you help start the workflow?  Thanks!

**Comment by delock** [comment]
@loadams the log shows I cannot execute conda install with sudo.   
```
$ sudo conda install -y -c conda-forge libstdcxx-ng=12
sudo: conda: command not found
```
However from a previous error log I can see conda is installed in /opt/conda and should be activated.
```
ImportError: /opt/conda/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /tmp/actions-runner/_work/DeepSpeed/DeepSpeed/oneCCL/build/_install/lib/libccl.so.1)
Error: Process completed with exit code 1.
```
So the question is whether there are suggestion how to install conda package in this SW configuration?

**Comment by delock** [comment]
Hi @loadams  I get suggestion from @1pikachu and @dbyoung18 that we can override the library file with `LD_PRELOAD` i.e.
```
export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
```
According to this link, we need to have libstdc++.so.6.0.30 or above to get GLIBCXX_3.4.30
https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html

So I would first probe which libstdc++ version is installed on the self-hosted system.

@1pikachu @dbyoung18, if you know better way to have this done please suggest.  Thanks!

> @loadams the log shows I cannot execute conda install with sudo.
> 
> ```
> $ sudo conda install -y -c conda-forge libstdcxx-ng=12
> sudo: conda: command not found
> ```
> 
> However from a previous error log I can see conda is installed in /opt/conda and should be activated.
> 
> ```
> ImportError: /opt/conda/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /tmp/actions-runner/_work/DeepSpeed/DeepSpeed/oneCCL/build/_install/lib/libccl.so.1)
> Error: Process completed with exit code 1.
> ```
> 
> So the question is whether there are suggestion how to install conda package in this SW configuration?

**Comment by loadams** [comment]
> Hi @loadams. I collected information from a ubuntu18.04 docker image. It turns out install gcc-9 would provide the system library we need. So LD_PRELOAD from system libstdc++.so.6 should give us the desired version. I made the change in workflow accordingly and can you help start the workflow? Thanks!

Apologies @delock - was out on vacation but it is running now.

**Comment by delock** [comment]
@loadams from the log, moving `source oneCCL/build/_install/env/setvars.sh` from unit test section to another section above cause the UT error.   Because communication ops needs to be built when the first time UT is running.  The latest PR add oneCCL env prep back.  Can you help restart the workflow?  Thanks!

**Comment by delock** [comment]
Hi @loadams, the failure in UT should be resolved by applying `export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libstdc++.so.6` in UT section as well.  Can you help restart the workflow?  Thanks!

**Comment by delock** [comment]
@loadams the CPU inference test has passed.  We also see tensor parallel test case bf16-marian has passed as expected.  Now we will turn on bf16-codegen.   This test could pass with pytorch 2.1.  After that this PR will be ready to merge and CPU inference workflow turned back on.

**Comment by delock** [comment]
the failure in CI is due to result mismatch from BF16 autoTP.  We have observed this on both CUDA device and CPU.   A probable reason is precision difference between BF16 and FP16.   We will disable codegen UT for BF16 to avoid UT failure.

**Comment by delock** [comment]
Hi @loadams this fix had passed all tests.   I think this PR is critical because all changes that affect CPU only should be guarded by this workflow.   Can we merge this workflow on master and turn back on CPU inference workflow?

**Comment by loadams** [comment]
> Hi @loadams this fix had passed all tests. I think this PR is critical because all changes that affect CPU only should be guarded by this workflow. Can we merge this workflow on master and turn back on CPU inference workflow? Thanks!

Hi @delock, yes apologies many of us were on vacation for the last week or so.  I've re-merged the master branch and we will get this reviewed/completed/merged and re-enable the cpu-inference tests ASAP.

**Comment by loadams** [comment]
@delock - looks like a failure on cpu-inference, I'll re-run to check that it isn't transient, but might need one more fix?

```
FAILED unit/inference/test_inference_config.py::TestInferenceConfig::test_overlap_kwargs - ValueError: Type fp16 is not supported.
FAILED unit/inference/test_inference_config.py::TestInferenceConfig::test_json_config - ValueError: Type fp16 is not supported.
```

**Comment by delock** [comment]
> @delock - looks like a failure on cpu-inference, I'll re-run to check that it isn't transient, but might need one more fix?
> 
> ```
> FAILED unit/inference/test_inference_config.py::TestInferenceConfig::test_overlap_kwargs - ValueError: Type fp16 is not supported.
> FAILED unit/inference/test_inference_config.py::TestInferenceConfig::test_json_config - ValueError: Type fp16 is not supported.
> ```

I think it's due to this recent change.  https://github.com/microsoft/DeepSpeed/pull/4843/files#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fR1020
I'll reproduce this locally and try to fix with a test update.

**Comment by delock** [comment]
@loadams I have reproduced these two failures locally.  The failing test is updated with setting dtype in config to float32, this should still test these two UTs as intended.

File: .github/workflows/cpu-inference.yml
Comment (loadams) [inline .github/workflows/cpu-inference.yml:59]: typo on troubleshooting*

File: deepspeed/comm/ccl.py
Comment (loadams) [inline deepspeed/comm/ccl.py:189]: What is the runtime error that we can hit here?

**Review by loadams — APPROVED** [review]

File: .github/workflows/cpu-inference.yml
Comment (delock) [reply .github/workflows/cpu-inference.yml:59]: Thanks!  Will fix it.

File: deepspeed/comm/ccl.py
Comment (delock) [reply deepspeed/comm/ccl.py:189]: The `while True:` loop would iterate rank and collect global rank until rank is out of local rank range.  In older version of PyTorch, this out-of-range will throw a ValueError.  In PyTorch 2, this behavior will throw a RuntimeError.

@Liangliang-Ma

File: tests/unit/inference/test_inference.py
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:75]: I think that caching the model list can be a good idea for the tests, but we need to save it to blob storage so that it is persistent. Additionally, I think the cache should have a timestamp connected to it, such that we update it every hour/day/week. See how we do this in MII:
https://github.com/microsoft/DeepSpeed-MII/blob/4472e4e206182ed56399f225848a7721565922fb/mii/utils.py#L39

File: tests/unit/inference/test_inference_config.py
Comment (mrwyattii) [inline tests/unit/inference/test_inference_config.py:18]: Why do we add dtype here?

Comment (mrwyattii) [reply tests/unit/inference/test_inference_config.py:18]: nevermind, I see the earlier conversation about this

**Review by mrwyattii — APPROVED** [review]
Given the long runtime of the cpu-inference tests, we should only require them on changes to `deepspeed/inference`.

Approving the PR so we can merge.

@loadams let's address the long test runtimes and the model list caching in a PR after this is merged.

**Pull request merged**
